File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters/DatabaseConnection.cs
// <copyright file="v.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters
{
    using Meta;

    public abstract class DatabaseConnection
    {
        protected DatabaseConnection(Configuration configuration) => this.Configuration = configuration;

        public Configuration Configuration { get; }

        public abstract IWorkspaceConnection CreateWorkspaceConnection();

        public abstract DatabaseRecord GetRecord(long id);

        public abstract long GetPermission(Class @class, IOperandType operandType, Operations operation);
    }
}

File: typescript/libs/workspace/system/adapters/src/lib/database/database-connection.ts
import {
  Configuration,
  IWorkspace,
  Operations,
} from '@allors/workspace/system/domain';
import { Class, OperandType } from '@allors/workspace/system/meta';

import { DefaultNumberRanges } from '../collections/ranges/default-number-ranges';
import { Ranges } from '../collections/ranges/ranges';
import { DatabaseRecord } from './database-record';

export type IdGenerator = () => number;

export abstract class DatabaseConnection {
  ranges: Ranges<number>;

  constructor(public configuration: Configuration) {
    this.ranges = new DefaultNumberRanges();
  }

  abstract createWorkspace(): IWorkspace;

  abstract getRecord(id: number): DatabaseRecord | undefined;

  abstract getPermission(
    cls: Class,
    operandType: OperandType,
    operation: Operations
  ): number | undefined;

  nextId(): number {
    return this.configuration.idGenerator();
  }
}
